bitbots_dynup/src/dynup_stabilizer.cpp

```diff
@@ -1,55 +1,81 @@
+  ros::NodeHandle nhr = ros::NodeHandle("dynup/pid_trunk_roll");
 
-  legs_joints_group_ = kinematic_model_->getJointModelGroup("Legs");
+  pid_trunk_pitch_.init(nhp, false);
+  pid_trunk_roll_.init(nhr, false);
 
   /* Reset kinematic goal to default */
   goal_state_.reset(new robot_state::RobotState(kinematic_model_));
   goal_state_->setToDefaultValues();
 }
 
 void Stabilizer::reset() {
-  /* We have to set some good initial position in the goal state,
-   * since we are using a gradient based method. Otherwise, the
-   * first step will be not correct */
-  std::vector<std::string> names_vec =
-      {"HeadPan", "HeadTilt", "LAnklePitch", "LAnkleRoll", "LElbow", "LHipPitch", "LHipRoll", "LHipYaw", "LKnee",
-       "LShoulderPitch", "LShoulderRoll", "RAnklePitch", "RAnkleRoll", "RElbow", "RHipPitch", "RHipRoll", "RHipYaw",
-       "RKnee", "RShoulderPitch", "RShoulderRoll"};
-  std::vector<double> pos_vec = {0, 47, -80, 0, -88, 65, -7, 2, -157, 11, -14, 80, 1, 97, -63, -3, 0, 157, -14, 14};
-  for (double &pos: pos_vec) {
-    pos = pos / 180.0 * M_PI;
-  }
-  for (int i = 0; i < names_vec.size(); i++) {
-    goal_state_->setJointPositions(names_vec[i], &pos_vec[i]);
-  }
+  pid_trunk_pitch_.reset();
+  pid_trunk_roll_.reset();
+
 }
 
-DynupResponse Stabilizer::stabilize(const DynupResponse &response, const ros::Duration &dt) {
-  // Currently, nothing happens here, but in the future, a PID controller for
-  // stabilizing or another stabilizing method could be implemented
-  return response;
+DynupResponse Stabilizer::stabilize(const DynupResponse &ik_goals, const ros::Duration &dt) {
+    tf2::Transform right_foot_goal;
+    tf2::Quaternion quat;
+    tf2::convert(imu_.orientation, quat);
+    // the tf2::Quaternion has a method to access roll pitch and yaw
+    double roll, pitch, yaw;
+    tf2::Matrix3x3(quat).getRPY(roll, pitch, yaw);
+
+    if(use_stabilizing_ && ik_goals.is_stabilizing_needed) {
+        tf2::Transform to_trunk_tf;
+        tf2::fromMsg(to_trunk_.transform, to_trunk_tf);
+        tf2::Transform trunk_goal = ik_goals.r_foot_goal_pose * to_trunk_tf;
+
+        double goal_pitch, goal_roll, goal_yaw;
+        tf2::Matrix3x3(trunk_goal.getRotation()).getRPY(goal_roll, goal_pitch, goal_yaw);
+        // first adapt trunk pitch value based on PID controller
+        double corrected_pitch = pid_trunk_pitch_.computeCommand(goal_pitch - pitch, dt);
+        double corrected_roll = pid_trunk_roll_.computeCommand(goal_roll - roll, dt);
+        tf2::Quaternion corrected_orientation;
+        corrected_orientation.setRPY(goal_roll + corrected_roll, goal_pitch + corrected_pitch, goal_yaw);
+        trunk_goal.setRotation(corrected_orientation);
+
+        right_foot_goal = trunk_goal * to_trunk_tf.inverse() ;
+    }
+    else {
+        right_foot_goal = ik_goals.r_foot_goal_pose;
+    }
+
+    tf2::Transform left_foot_goal = ik_goals.l_foot_goal_pose * right_foot_goal;
```

Comment:
I don't understand how this works. My understanding is that `right_foot_goal` is the – possibly stabilized – goal pose of the right foot, relative to the `base_link`. `ik_goals.l_foot_goals` is the (unstabilized) pose of the left foot, also relative to `base_link`. I assume the goal of this line is to apply the stabilization of the right foot to the left foot. But multiplying the frames would result in a pose that is not the goal pose of the left foot. Since the code is working, I have probably a bad understanding of the frames in which the poses are present. Generally, that is an information that I am missing in the code – which pose is in which frame? I would like to have more clarifying comments.

Reply:
Okay, after reading `dynup_engine.cpp`, I understand that `left_foot_goal` is always relative to the right foot. But it should be documented in a central place, for example in the definitions of `DynupResponse` and `DynupRequest` in `dynup_utils.cpp`, as well as in the general documentation on how the splines are created.